### PR TITLE
Review Code, Add Tests and BinghamMetric Class instance, and first version of tutorial.

### PR DESCRIPTION
**Pull request by RafaelNH**
Hi @CHrlS98,

Here is my input for the Bingham fitting functions. 

What I've done:

1. I reviewed all functions and improved their documentation

2. As I promised, I added the test
3. As I also promised, I added the BinghamMetric Class instance in the function to the fit Binghams in multiple voxels - to be more efficient in dealing with different Bingham metrics. 
4. Created a first version of the tutorial with the description of some of the Bingham metrics.

Two notes:

- We may consider increasing the default value for max_search_angle. the tests were not passing with the current default value. Indeed, you can notice that in the test I am using a larger value for max_search_angle (45 degrees)
- (this may be a comment for Julio) Regarding the ´bingham_from_sh´ function, it is not recommended to load data and processes computation in the same function. Therefore, I've developed a new function named ´bingham_from_odf´ to perform Bingham fitting directly from ODF maps. I understand there might be concerns regarding the memory demands of handling ODF maps directly. To address this, we could consider offering an alternative function that operates on spherical harmonics (that should be inputted not loaded). If you believe incorporating spherical harmonics input would be beneficial, please feel free to adjust ´bingham_from_odf´ accordingly and update the example.

### CODE REVIEW

**Comment by villalonreina** [comment]
@RafaelNH Thanks so much for all this work. Yes, I agree that it's better to fit Bingham directly on the ODF maps. It makes sense to have the `bingham_from_odf` function instead of `bingham_from_sh`. Maybe we can have one example/tutorial where a spherical harmonics file is inputted, then converted to odf (with `sh_to_sf`), and then call `bingham_from_odf`. What do you think? The thing is that I think that in practice most researchers will fit their ODFs for other purposes (tractography most likely) and will save the spherical harmonics to disk (likely in Nifti files), which then could be reused to fit the Bingham metrics.

**Comment by RafaelNH** [comment]
@CHrlS98 As you requested I rebased the code according to your latest branch version. 

@villalonreina I added now the function to fit "Binghams" from spherical harmonics and include some notes in documentation.

I think I am done here. Feel free to review the code and merge it to DIPY's PR if you don't detect any issues. After that we can notify the DIPY team that this is ready to be merged and ask for their input.

File: dipy/direction/bingham.py [hunk 1]
```diff
@@ -115,134 +55,279 @@ def _bingham_fit_peak(sf, peak, sphere, max_angle):
...
+    The Bingham function is defined as [1]_,[2]_,[3]_:
+
+    .. math::
+
+        f(n) = f_0 \exp[-k_1(\mu_1^Tn)^2-k_2(\mu_2^Tn)^2]
+
+    where $f(n)$ is the Bingham function value at a given direction $n$, $f_0%
+    is the Bingham maximum peak value, $k_1$ and $k_2$ are the concentration
+    constants parameters (large values correspond to lower dispersion values)
+    along the two dispersion axes $\mu_1$ and $\mu_2$-
```
Comment (CHrlS98) [inline dipy/direction/bingham.py]:
```suggestion
    where $f(n)$ is the Bingham function value at a given direction $n$, $f_0$
    is the Bingham maximum peak value, $k_1$ and $k_2$ are the concentration
    constants parameters (large values correspond to lower dispersion values)
    along the two dispersion axes $\mu_1$ and $\mu_2$.
```

File: dipy/direction/bingham.py [hunk 2]
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
...
 
 
-def bingham_from_sh(sh, mask, sh_order, npeaks, sphere, max_angle,
-                    min_sep_angle, rel_th):
-    r"""
-    Function that calls in the 4D Spherical Harmonics file of a brain image
-    and loops through the image to fit the Bingham distribution at each voxel.
+def odi2k(odi):
+    """ Convent Bingham/Watson concentration parameter k to the orientation
+    dispersion index (ODI)
```
Comment (CHrlS98) [inline dipy/direction/bingham.py]:
```suggestion
    """ Convert Bingham/Watson concentration parameter k to the orientation
    dispersion index (ODI)
```

File: dipy/direction/bingham.py [hunk 3]
```diff
@@ -115,134 +55,279 @@ def _bingham_fit_peak(sf, peak, sphere, max_angle):
...
-        Bingham distributions. Each tuple describes a lobe of the
-        initially fitted function.
+    bingham_params : ndarray (...., nl, 12)
+        ndarray containing the model parameters of Binghams fitted to ODFs in
+        the following order:
+            Maximum value of the Bingham function (f0, index 0);
+            concentration parameters k1 and k2 (indexes 1 and 2);
+            elements of Bingham main direction (indexes 3-5);
+            elements of Bingham dispersion major axis (indexes 6-8);
+            and elements of Bingham dispersion minor axis (indexs 9-11).
```
Comment (CHrlS98) [inline dipy/direction/bingham.py]:
```suggestion
            and elements of Bingham dispersion minor axis (indexes 9-11).
```

File: dipy/direction/bingham.py [hunk 4]
```diff
@@ -115,134 +55,279 @@ def _bingham_fit_peak(sf, peak, sphere, max_angle):
...
+    Parameters
+    ----------
+    bingham_params : ndarray (...., nl, 12)
+        ndarray containing the model parameters of Binghams fitted to ODFs in
+        the following order:
+            Maximum value of the Bingham function (f0, index 0);
+            concentration parameters k1 and k2 (indexes 1 and 2);
+            elements of Bingham main direction (indexes 3-5);
+            elements of Bingham dispersion major axis (indexes 6-8);
+            and elements of Bingham dispersion minor axis (indexs 9-11).
```
Comment (CHrlS98) [inline dipy/direction/bingham.py]:
```suggestion
            and elements of Bingham dispersion minor axis (indexes 9-11).
```

File: dipy/direction/bingham.py [hunk 5]
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
...
+    return fs
 
 
-def bingham_orientation_dispersion(bingham_fits):
-    r"""
-    Compute the orientation dispersion indexes (ODI) from the
-    concentration parameters (k1, k2) as described in [2]_ and [3]_.
-    
+def k2odi(k):
+    """ Convent Bingham/Watson concentration parameter k to the orientation
```
Comment (CHrlS98) [inline dipy/direction/bingham.py]:
```suggestion
    """ Convert Bingham/Watson concentration parameter k to the orientation
```

File: dipy/direction/bingham.py [hunk 6]
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
...
+def _convert_bingham_pars(fits, npeaks):
+    """ Convert tuple output of Bingham fit to ndarray
+
+    Parameters
+    ----------
+    fits : tuple
+       tuple of nl elements containting the Bingham function parameters
+       in the following order:
+            Maximum value of the Bingham function (f0);
+            concentration parameters (k1 and k2;
```
Comment (CHrlS98) [inline dipy/direction/bingham.py]:
```suggestion
            concentration parameters (k1 and k2);
```

File: dipy/direction/bingham.py [hunk 7]
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
...
+    Returns
+    -------
+    bingham_params : ndarray (nl, 12)
+        ndarray containing the model parameters of Binghams fitted to ODFs in
+        the following order:
+            Maximum value of the Bingham function (f0, index 0);
+            concentration parameters k1 and k2 (indexes 1 and 2);
+            elements of Bingham main direction (indexes 3-5);
+            elements of Bingham dispersion major axis (indexes 6-8);
+            and elements of Bingham dispersion minor axis (indexs 9-11).
```
Comment (CHrlS98) [inline dipy/direction/bingham.py]:
```suggestion
            and elements of Bingham dispersion minor axis (indexes 9-11).
```

File: dipy/direction/bingham.py [hunk 8]
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
...
+        bpars[ln, 2] = fits[ln][2]
+        bpars[ln, 3:6] = fits[ln][3]
+        bpars[ln, 6:9] = fits[ln][4]
+        bpars[ln, 9:12] = fits[ln][5]
+    return bpars
+
+
+def global_voxel_metric(bmetric, bfd):
+    """ Compute a global scalar maps for metrics of Bingham functions
+    fitted to multiple ODF lobel
```
Comment (CHrlS98) [inline dipy/direction/bingham.py]:
```suggestion
    fitted to multiple ODF lobes
```

File: dipy/direction/bingham.py [hunk 9]
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
...
+        Parameters
+        ----------
+        model_params : ndarray (..., nl, 12)
+            ndarray containing the model parameters of Binghams fitted to ODFs
+            in the following order:
+            Maximum value of the Bingham function (f0, index 0);
+            concentration parameters k1 and k2 (indexes 1 and 2);
+            elements of Bingham main direction (indexes 3-5);
+            elements of Bingham dispersion major axis (indexes 6-8);
+            and elements of Bingham dispersion minor axis (indexs 9-11).
```
Comment (CHrlS98) [inline dipy/direction/bingham.py]:
```suggestion
            and elements of Bingham dispersion minor axis (indexes 9-11).
```

File: dipy/direction/bingham.py [hunk 10]
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
...
+        return self.model_params[..., 1]
+
+    @auto_attr
+    def kappa_2(self):
+        """Concentration parameter k2 for each ODF lobe."""
+        return self.model_params[..., 2]
+
+    @auto_attr
+    def kappa_total(self):
+        """Comnbined concentration parameters for each ODF lobe.
```
Comment (CHrlS98) [inline dipy/direction/bingham.py]:
```suggestion
        """Combined concentration parameters for each ODF lobe.
```

File: dipy/direction/bingham.py [hunk 11]
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
...
+
+        .. math::
+            k_{total} = sqrt{(k_1k_2)}
+        """
+        return np.sqrt(self.kappa_1 * self.kappa_2)
+
+    @auto_attr
+    def odi_1(self):
+        """Orientation Dispersion index 1 computed for each ODF lobe from
+        concentration paramenter kappa_1."""
```
Comment (CHrlS98) [inline dipy/direction/bingham.py]:
```suggestion
        concentration parameter kappa_1."""
```

File: dipy/direction/bingham.py [hunk 12]
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
...
+        return k2odi(self.kappa_1)
+
+    @auto_attr
+    def odi_2(self):
+        """Orientation Dispersion index 1 computed for each ODF lobe from
+        concentration paramenter kappa_2."""
```
Comment (CHrlS98) [inline dipy/direction/bingham.py]:
```suggestion
        concentration parameter kappa_2."""
```

File: doc/examples/reconst_bingham.py [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+# the Axis with Lesser Dispersion (directions in which fiber orientations are
+# more uniform); and 3) Average ODI of the second-largest ODF lobe across noth
+# axes. In this figure, regions of white matter that contain only a single
+# fiber  population display ODI estimates of zero, corresponding to ODF
+# profiles lacking a second ODF lobe.
+#
+# BinghamMetric can also be used to compute the averaged ODI quantities across
+# all ODF lobes (see below). The averaged quantitaties are computed by
+# weigthing each ODF lobe with their respective FD value. These quantities
+# are ploted in the following figure.
```
Comment (CHrlS98) [inline doc/examples/reconst_bingham.py]:
```suggestion
# are plotted in the following figure.
```

**Review by CHrlS98 — COMMENTED** [review]
I read the comments and tutorial and noted some typos. I added them as suggestions. Otherwise it looks good to me, thanks a lot!

**Comment by RafaelNH** [comment]
Thanks for spotting these typos. I've incorporated all of them

**Comment by CHrlS98** [comment]
Thank you, I'll go ahead with the merge.

**Comment by villalonreina** [comment]
@CHrlS98 Is it merged already?

**Comment by CHrlS98** [comment]
@villalonreina yes, you can review the DIPY PR directly: https://github.com/dipy/dipy/pull/2826

**Pull request merged**